internal/outctx/manager.go

```diff
@@ -19,38 +20,164 @@ import (
 const s3TagKey = "fluentBitTag"
 
 // Resources and metadata to process Fluent Bit events with the same tag.
-type EventManager struct {
-	Tag    string
-	Index  int
-	Writer irzstd.Writer
+type S3EventManager struct {
+	Tag            string
+	Index          int
+	Writer         irzstd.Writer
+	Mutex          sync.Mutex
+	WaitGroup      sync.WaitGroup
+	UploadRequests chan bool
+	Listening      bool
 }
 
-// Sends Zstd buffer to s3 and reset writer and buffers for future uploads. Prior to upload,
-// IR buffer is flushed and IR/Zstd streams are terminated. The [EventManager.Index] is incremented
-// on successful upload.
+// Ends listener goroutine.
+func (m *S3EventManager) StopListening() {
+
+	if !m.Listening {
+		return
+	}
+
+	log.Printf("Stopping upload listener for event manager with tag %s", m.Tag)
+
+	// Closing the channel sends terminate signal to goroutine. The WaitGroup
+	// will block until it actually terminates.
+	close(m.UploadRequests)
+	m.WaitGroup.Wait()
+	m.Listening = false
+}
+
+// Starts upload listener which can receive signals on UploadRequests channel. This function should
+// be called as a goroutine. Timeout is only triggered if use_disk_buffer is on. Function calls
+// immortal functions and thus will not exit immediately. Instead, it will only exit if the
+// uploadRequest channel is closed which will allow immortal functions to break out of infinite
+// loop. When function does exit, it decrements a WaitGroup letting event manager know it has
+// exited. WaitGroup allows graceful exit of listener when Fluent Bit receives a kill signal. On
+// [recovery.GracefulExit], plugin will wait to exit until all listeners are closed. Without
+// WaitGroup, OS may abruptly kill listen goroutine.
 //
 // Parameters:
 //   - config: Plugin configuration
 //   - uploader: S3 uploader manager
+func (m *S3EventManager) listen(config S3Config, uploader *manager.Uploader) {
+	defer m.WaitGroup.Done()
+
+	m.Listening = true
+	if m.Writer.GetUseDiskBuffer() {
+		m.diskUploadListener(config, uploader)
+	} else {
+		m.memoryUploadListener(config, uploader)
+	}
+}
```

Comment:
Lets refactor this. I could be wrong, but it looks like all of the upload path should be merged. We should still have a timeout in the memory buffered version and we can call `CheckEmpty` which should cover the differences.

Essentially, we shouldn't need to know if it is memory or disk buffering in the manager code.